### PR TITLE
chore: respond step name for waiting

### PR DESCRIPTION
**Pull request by leangseu-edx**
**TL;DR -** Handle waiting state

JIRA: https://2u-internal.atlassian.net/browse/AU-1542

**What changed?**

- On waiting with `staff` required, return staff
- On waiting with `peer`, return peer
- (Unit test) If staff not require, it should be done

**Developer Checklist**

- [ ] Reviewed the [release process](https://github.com/openedx/edx-ora2/blob/master/.github/release_process.md)
- [ ] Translations and JS/SASS compiled
- [ ] Bumped version number in [openassessment/\_\_init\_\_.py](https://github.com/openedx/edx-ora2/blob/master/openassessment/__init__.py#L4) and [package.json](https://github.com/openedx/edx-ora2/blob/master/package.json#L3)

**Testing Instructions**

[ How should a reviewer test this PR? ]

**Reviewer Checklist**

Collectively, these should be completed by reviewers of this PR:

- [ ] I've done a visual code review
- [ ] I've tested the new functionality

FYI: @openedx/content-aurora

### CODE REVIEW

File: openassessment/xblock/ui_mixins/mfe/page_context_serializer.py [hunk 1]
```diff
@@ -248,7 +252,16 @@ def get_activeStepName(self, instance):
         if not instance.workflow_data.has_workflow:
             return "submission"
         else:
-            return STEP_NAME_MAPPINGS[instance.workflow_data.status]
+            step_name = STEP_NAME_MAPPINGS[instance.workflow_data.status]
```
Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/page_context_serializer.py:255]: How about we remove waiting from step_name_mappings and then do

```
elif instance.workflow_data.is_waiting:
    <what you have here>
else:
     return STEP_NAME_MAPPINGS[instance.workflow_data.status]

```

File: openassessment/xblock/ui_mixins/mfe/test_page_context_serializer.py [hunk 2]
```diff
@@ -606,6 +619,29 @@ def test_self_assessment_not_available(self, xblock):
 
         self.assertNestedDictEquals(expected_data, progress_data)
 
+    @scenario('data/grade_scenario_self_staff.xml', user_id='Alan')
+    def test_self_staff_assessment__waiting(self, xblock):
+        # Given I am on the staff step and waiting for submissions
+        self.create_submission_and_assessments(xblock, 'submission_text', [], [], SELF_ASSESSMENT)
+        xblock.workflow_data.workflow['status'] = 'waiting'
```
Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/test_page_context_serializer.py:626]: I don't think this line is doing anything. Did you mean for this to be
```suggestion
        assert xblock.workflow_data.workflow['status'] == 'waiting'
```

File: openassessment/xblock/ui_mixins/mfe/test_page_context_serializer.py [hunk 3]
```diff
@@ -606,6 +619,29 @@ def test_self_assessment_not_available(self, xblock):
...
+
+    @scenario('data/grade_scenario_self_staff_not_required.xml', user_id='Alan')
+    def test_self_staff_assessment__not_required(self, xblock):
+        # Given I am on the staff step and waiting for submissions
+        self.create_submission_and_assessments(xblock, 'submission_text', [], [], SELF_ASSESSMENT)
+
+        # When I ask for progress
+        progress_data = ProgressSerializer(xblock).data
+
+        # Expect active step to be staff instead of waiting
```
Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/test_page_context_serializer.py:642]: This comment doesn't reflect what's happening. I think this test is actually not necessary because as soon as we submit the self assessment in this case, since the staff is not required (and therefore auto-generated) it's just skipped and we go straight to done rather than waiting

File: openassessment/xblock/ui_mixins/mfe/test_page_context_serializer.py [hunk 4]
```diff
@@ -475,6 +475,19 @@ def test_peer_assessment(self, xblock):
 
         self.assertNestedDictEquals(expected_data, progress_data)
 
+    @scenario("data/peer_only_scenario.xml", user_id="Alan")
+    def test_peer_assessment__waiting(self, xblock):
+        # Given I am on the peer step and waiting for submissions
+        self.create_test_submission(xblock)
+        # xblock.workflow_data.workflow.set("status", "waiting")
+        xblock.workflow_data.workflow["status"] = "waiting"
```
Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/test_page_context_serializer.py:483]:
```suggestion
        # The block requires us to grade 5, and be graded by three, so have us grade 5, and be graded by none
        peers = ['p1', 'p2', 'p3', 'p4', 'p5']
        self.create_submission_and_assessments(xblock, 'submission_text', peers, [self.PEER_ASSESSMENTS[0]] * 5, None, waiting_for_peer=True)
        assert xblock.workflow_data.workflow["status"] = "waiting"
```

File: openassessment/xblock/ui_mixins/mfe/test_page_context_serializer.py [hunk 5]
```diff
@@ -606,6 +619,29 @@ def test_self_assessment_not_available(self, xblock):
...
+    @scenario('data/grade_scenario_self_staff_not_required.xml', user_id='Alan')
+    def test_self_staff_assessment__not_required(self, xblock):
+        # Given I am on the staff step and waiting for submissions
+        self.create_submission_and_assessments(xblock, 'submission_text', [], [], SELF_ASSESSMENT)
+
+        # When I ask for progress
+        progress_data = ProgressSerializer(xblock).data
+
+        # Expect active step to be staff instead of waiting
+        self.assertEqual('done', progress_data['activeStepName'])
```
Comment (nsprenkle) [inline openassessment/xblock/ui_mixins/mfe/test_page_context_serializer.py:643]: Your comment contradicts the actual test. Which is correct?

File: openassessment/xblock/ui_mixins/mfe/page_context_serializer.py [hunk 6]
```diff
@@ -248,7 +252,16 @@ def get_activeStepName(self, instance):
...
+            if step_name == "waiting":
+                workflow_requirements = instance.workflow_data.workflow_requirements
+                if 'staff' in workflow_requirements and workflow_requirements['staff']['required']:
+                    return "staff"
+                elif 'peer' in workflow_requirements:
+                    return "peer"
+                else:
+                    raise UnknownActiveStepException("Workflow is in waiting but no staff or peer step is required.")
+            return step_name
```
Comment (nsprenkle) [inline openassessment/xblock/ui_mixins/mfe/page_context_serializer.py:264]: Nit: Can remove the need to early assign step name, and use more consistent quotation marks.

```suggestion
            if instance.workflow_data.status == "waiting":
                workflow_requirements = instance.workflow_data.workflow_requirements
                if "staff" in workflow_requirements and workflow_requirements["staff"]["required"]:
                    return "staff"
                elif "peer" in workflow_requirements:
                    return "peer"
                else:
                    raise UnknownActiveStepException("Workflow is in waiting but no staff or peer step is required.")
            return STEP_NAME_MAPPINGS[instance.workflow_data.status]
```

**Review by nsprenkle — COMMENTED** [review]
Open question on one of the tests

File: openassessment/xblock/ui_mixins/mfe/test_page_context_serializer.py [hunk 7]
```diff
@@ -606,6 +619,29 @@ def test_self_assessment_not_available(self, xblock):
...
+
```
Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/test_page_context_serializer.py:644]: Add an xml test scenario for peer then staff, then write a test with the same setup as my recommended setup for your peer step above, where we're waiting on peer AND staff, and assert that the active step will be "staff"

Comment (jansenk) [inline openassessment/xblock/ui_mixins/mfe/test_page_context_serializer.py:644]: Add a test case where we are in a self-only ORA (no staff or peer), then patch the workflow API object's is_waiting to return True. Assert the expected error is raised.

**Review by jansenk — CHANGES_REQUESTED** [review]
Some test changes

**Review by jansenk — APPROVED** [review]
